src/org/sosy_lab/java_smt/basicimpl/IndependentInterpolatingProverEnvironment.java

```diff
@@ -0,0 +1,519 @@
+import org.sosy_lab.java_smt.api.SolverContext.ProverOptions;
+import org.sosy_lab.java_smt.api.SolverException;
+import org.sosy_lab.java_smt.api.UFManager;
+
+// The non-interpolating delegates usually return nothing useful as itp points, so we have to
+// track it here
+@SuppressWarnings("unused")
+public class IndependentInterpolatingProverEnvironment<TFormulaInfo, TType>
+    extends AbstractProver<TFormulaInfo> implements InterpolatingProverEnvironment<TFormulaInfo> {
+
+  private final SolverContext solverContext;
+  private final ShutdownNotifier shutdownNotifier;
+  private final ProverEnvironment delegate;
+
+  private final FormulaCreator<TFormulaInfo, TType, ?, ?> creator;
+  private final FormulaManager mgr;
+  private final BooleanFormulaManager bmgr;
+  private final UFManager ufmgr;
+  private final QuantifiedFormulaManager qfmgr;
+
+  // null for native solver interpolation
+  private final @Nullable ProverOptions interpolationStrategy;
+
+  private static final UniqueIdGenerator UNIQUE_ID_GENERATOR = new UniqueIdGenerator();
+  private static final String PREFIX = "__internal_model_itp_generation_";
+
+  public IndependentInterpolatingProverEnvironment(
+      SolverContext pSourceContext,
+      FormulaCreator<TFormulaInfo, TType, ?, ?> pCreator,
+      ProverEnvironment pDelegate,
+      Set<ProverOptions> pOptions,
+      ShutdownNotifier pShutdownNotifier) {
+    super(pOptions);
+    solverContext = checkNotNull(pSourceContext);
+    delegate = checkNotNull(pDelegate);
+    creator = pCreator;
+    shutdownNotifier = pShutdownNotifier;
+    interpolationStrategy = getIndependentInterpolationStrategy(pOptions);
+    mgr = pSourceContext.getFormulaManager();
+    bmgr = mgr.getBooleanFormulaManager();
+    ufmgr = mgr.getUFManager();
+    qfmgr = mgr.getQuantifiedFormulaManager();
+  }
+
+  /**
+   * Checks, whether an independent interpolation strategy is enabled for the given prover.
+   *
+   * @param options A set of all enabled options for the prover to check.
+   * @return {@code true} if an independent interpolation strategy is configured, {@code false}
+   *     otherwise.
+   */
+  public static boolean hasIndependentInterpolationStrategy(Set<ProverOptions> options) {
+    return getIndependentInterpolationStrategy(options) != null;
+  }
+
+  private static @Nullable ProverOptions getIndependentInterpolationStrategy(
+      Set<ProverOptions> options) {
+    List<ProverOptions> itpStrat = new ArrayList<>(options);
+    final Set<ProverOptions> allIndependentInterpolationStrategies =
+        ImmutableSet.of(
+            ProverOptions.GENERATE_MODEL_BASED_INTERPOLANTS,
+            ProverOptions.GENERATE_UNIFORM_BACKWARD_INTERPOLANTS,
+            ProverOptions.GENERATE_UNIFORM_FORWARD_INTERPOLANTS);
+    itpStrat.retainAll(allIndependentInterpolationStrategies);
+    if (itpStrat.isEmpty()) {
+      return null;
+    }
+    Preconditions.checkState(itpStrat.size() == 1);
```

Comment:
Since we only allow exactly 1 solver independent interpolation option, we could fail here more gracefully with an `InvalidConfigurationException` explaining the problem.